### PR TITLE
Add fish_trace_depth

### LINKED ISSUE #9069: Request: Fish Trace depth
fish, version 3.5.0

Would be amazing if e.g. `set fish_trace 1` would only print first-level traces (one indentation level), as I often only want to see the commands a script executes, but not everything inside them.

As this would be backwards-incompatible I thought instead to add a `fish_trace_depth` variable.

### PR DESCRIPTION
**Pull request by tjkirch**
This allows limiting the depth to which fish_trace will trace execution. For example, setting it to 0 will only print top-level commands, a value of 1 will print commands inside a for loop from the top level, etc.

Fixes #9069

## TODOs:

- [x] If addressing an issue, a commit message mentions `Fixes issue #<issue-number>`
- [x] Changes to fish usage are reflected in user documentation/manpages.
- [x] Tests have been added for regressions fixed
- [x] User-visible changes noted in CHANGELOG.rst

### CODE REVIEW

File: tests/checks/trace.fish [hunk 1]
```diff
@@ -65,6 +65,69 @@ end
...
+end
+
+# CHECK: 1 a
+# CHECK: 1 b
+# CHECK: 2 a
+# CHECK: 2 b
+
+# CHECKERR: > for 1 2
+# CHECKERR: -> for a b
+# CHECKERR: --> echo 1 a
```
Comment (krobelus) [inline tests/checks/trace.fish:87]: this is less discoverable than grep, so maybe not worth it

Comment (tjkirch) [reply tests/checks/trace.fish:87]: I think it's simpler and more discoverable in at least two cases:
1. I'm running a number of commands interactively that I'm trying to understand better, and I don't have to think about how or where to add grep on each one.
2. I'm writing a script.  I want to debug by showing what commands are being run without adding a lot of `echo`s, and I don't want to need `exec` tricks, extra scopes, or a wrapper script for grep.

In other words, `fish_trace_depth` is useful in more complex scenarios, which is exactly where you might want `fish_trace` in the first place.

In both cases, it's helpful to see a level or two of depth, but not the full trace that shows complete function expansions, etc., so it's convenient to have this one obvious setting.

Comment (Nahor) [reply tests/checks/trace.fish:87]: Other tests:
- set the depth based on specific context (e.g. only deep traces for i=2, don't care about i=1 or i=3 so shallow traces)
- setting a deep depth after going beyond the current max, i.e. a discontinuity in the depth (e.g. start with a depth of 1, then when at the script is a depth 3, set the depth to 4)
- set the depth in a local scope (variant of the specific context above, but with automatic reset)

Comment (tjkirch) [reply tests/checks/trace.fish:87]: @Nahor I've pushed an update that hopefully addresses your requests.  Let me know if I didn't understand and I can adjust.

File: tests/checks/trace.fish [hunk 2]
```diff
@@ -65,6 +65,199 @@ end
...
+
+# Nothing inside i=3 because depth returns to 0 automatically
+
+# Now, test increasing the depth after exceeding some limit where we would start
+# caring more about the detailed trace.
+
+set fish_trace_depth 0
+# CHECKERR: > set fish_trace_depth 0
+
+for i in 1 2 3
```
Comment (Nahor) [inline tests/checks/trace.fish]: You don't need a for loop for this one, maybe something like:
```fish
set fish_trace_depth 1
if true
  if true
    if true
      set -l fish_trace_depth 10
      if true
        true
      end
    end      
  end
end
```
This should make the `# CHECKERR` more obvious and self-explanatory
Using `true` instead of `echo` also avoids useless `# CHECK` on stdout.

Also, yours is not deep enough. The idea is to check the discontinuity in the output, e.g. `->` then `--->`, no `-->` in between.

File: src/env_dispatch.rs [hunk 3]
```diff
@@ -347,6 +348,21 @@ fn handle_fish_trace(vars: &EnvStack) {
     );
 }
 
+fn handle_fish_trace_depth(vars: &EnvStack) {
+    if let Some(depth_str) = vars.get(L!("fish_trace_depth")) {
```
Comment (Nahor) [inline src/env_dispatch.rs]: It would be cleaner to do:
```rust
let Some(depth_str) = vars.get(...) else {
   flog!(...);
   return;
}
let Ok(new_depth = fish_wcstoi(...) else {
   flog!(...);
   return;
}
...
```
It's more sequential, less nesting.

File: src/env_dispatch.rs [hunk 4]
```diff
@@ -347,6 +348,21 @@ fn handle_fish_trace(vars: &EnvStack) {
...
+        if let Ok(new_depth) = fish_wcstoi(&depth_str.as_string()) {
+            if let Ok(new_depth) = usize::try_from(new_depth) {
+                crate::trace::trace_set_depth(new_depth);
+                flog!(term_support, "Setting trace depth to ", new_depth);
+            } else {
+                flog!(term_support, "Trace depth needs to be a positive integer");
+            }
+        } else {
+            flog!(term_support, "Trace depth needs to be an integer");
```
Comment (Nahor) [inline src/env_dispatch.rs]: 9999999999999999999999 is an integer but will fail the function (overflow)

File: tests/checks/trace.fish [hunk 5]
```diff
@@ -65,6 +65,199 @@ end
...
+# CHECKERR: --> set fish_trace_depth 10
+# CHECKERR: -> end if
+# CHECKERR: -> for a b
+# CHECKERR: --> echo 3 a
+# CHECKERR: --> echo 3 b
+# CHECKERR: -> end for
+# CHECKERR: > end for
+
+# Finally, make sure we can end tracing and see no more stderr.
+
```
Comment (Nahor) [inline tests/checks/trace.fish:281]: It's also missing tests for invalid depth values (not a number, negative, overflow, empty, ...).

File: src/env_dispatch.rs [hunk 6]
```diff
@@ -347,6 +348,21 @@ fn handle_fish_trace(vars: &EnvStack) {
     );
 }
 
+fn handle_fish_trace_depth(vars: &EnvStack) {
+    if let Some(depth_str) = vars.get(L!("fish_trace_depth")) {
+        if let Ok(new_depth) = fish_wcstoi(&depth_str.as_string()) {
+            if let Ok(new_depth) = usize::try_from(new_depth) {
+                crate::trace::trace_set_depth(new_depth);
+                flog!(term_support, "Setting trace depth to ", new_depth);
```
Comment (Nahor) [inline src/env_dispatch.rs]: Wrong category, should be `env_dispatch` I believe.

File: doc_src/language.rst [hunk 7]
```diff
@@ -1630,6 +1630,10 @@ You can change the settings of fish by changing the values of certain variables.
    It goes to stderr by default.
    Set it to ``all`` to also trace execution of key bindings, event handlers as well as prompt and title functions.
 
+.. envvar:: fish_trace_depth
+
+   if set to a nonnegative integer, will limit the number of levels printed by :envvar:`fish_trace`.
```
Comment (Nahor) [inline doc_src/language.rst:1635]: It should specify that the "first" level is "0".

And for that matter should it be "0"? or "1"? fish array indices are 1-based. The variable description also says "how many level", in which case "1" should be the first. And often enough for this kind of things, "0" is used to mean "disabled", aka "0 level shown". On the other hand, tracing is for script developers, and developers are used to 0-based, and the number matches the number of `-` in the `--->` output (but then, for 1-based one could say it's the number of characters instead of just `-` so...🤷).

**Review by Nahor — COMMENTED** [review]
Full review this time (before it was just a comment on top of my head while I was browsing the pending PRs 😛)

File: tests/checks/trace.fish (hunk 2 above)
Comment (tjkirch) [reply tests/checks/trace.fish]: Thanks, this helped explain what you were looking for, and it did clean things up.

File: src/env_dispatch.rs (hunk 4 above)
Comment (tjkirch) [reply src/env_dispatch.rs]: I agree that it's not 100% precise, but I wasn't sure it was helpful to expand it to something like "Trace depth needs to be a positive integer _that fits within the pointer size of your architecture_" or anything like that.  Am I misunderstanding?  Is there a simple, more precise way to say it?

File: tests/checks/trace.fish (hunk 5 above)
Comment (tjkirch) [reply tests/checks/trace.fish:281]: Thanks.  Adding these helped me find a bug: `handle_fish_trace_depth` needed to reset the depth to the default when the variable is unset.  Clearly if someone does `set -e fish_trace_depth` they want the default again.

However, it did raise another question.  If trace depth is set to 1, for example, and then the user tries to set it to "a" or another invalid value, should it stay set to 1 internally, or should it be reset to the default?  My thought is that it should be reset to the default, because otherwise the user could be confused when they're seeing a tracing depth of 1 when fish_trace_depth is set to "abc".  (In other words, I think the behavior of fish_trace_depth=abc should be the same regardless of what happened to the variable earlier.)  I believe this matches the behavior of fish elsewhere; for example, if I set a fish_color_* variable to nonsense, things just show up as white (a default color), not as the previous color they were set to.

This also helped me realize that one of my tests was bad.  If the depth variable is empty, it will become the default value (see above), which means that shadowing it with a smaller scope and a conditional `set` will briefly use the default trace depth.  This was the test code:
```fish
set fish_trace_depth 0

for i in 1 2 3
    set --local fish_trace_depth
    if test "$i" -eq 2
        set fish_trace_depth 10
    end

    ...things to trace at depth 10 here...
```

As soon as the `set --local fish_trace_depth` runs, depth will reset to the default, because it has nothing else to go on.  Then if the conditional passes, it will be set to 10.  In between the two, trace depth is max, so everything would be shown, but that's not what I had in CHECKERR, because I hadn't thought about this case.

The test you suggested, which I added right after this one, is a better test of setting depth in local scope, so I deleted this test.

File: src/env_dispatch.rs (hunk 6 above)
Comment (tjkirch) [reply src/env_dispatch.rs]: I had originally copied `handle_emoji_width` for this bit, which uses `term_support`, and given your message, I think it may be incorrect there as well.

File: doc_src/language.rst (hunk 7 above)
Comment (tjkirch) [reply doc_src/language.rst:1635]: I added that line to the docs.

And yeah, this could be argued either way.  I went with 0 meaning "only show the top level" because then 1 means "show 1 level of depth".  I wouldn't count the top-level commands as having a level of depth; to me, that would mean something inside a scope.  But obviously this can be argued either way, so I'm happy to change it if that's what people think is best.

File: tests/checks/trace.fish [hunk 8]
```diff
@@ -65,6 +65,239 @@ end
...
+# CHECKERR: --> set fish_trace_depth 0
+# CHECKERR: > end for
+
+# Now, test increasing the depth after exceeding the previous depth. We set
+# depth in a local scope so it's automatically reset.
+
+set fish_trace_depth 1
+# CHECKERR: > set fish_trace_depth 1
+
+if true
```
Comment (Nahor) [inline tests/checks/trace.fish]: Don't know if it's worth changing, but `{ ... }` or `begin; ...; end` is even better than my previous suggestion, because it's only one trace (`begin`) instead of two (`if+true`), and the fish code could be on a single line (`{ { { true } } }`, looks like fish_indent is OK with that. see another example in a comment below)

File: tests/checks/trace.fish [hunk 9]
```diff
@@ -65,6 +65,239 @@ end
...
+# The locally set depth of 10 goes out of scope here, so we don't see the
+# intermediate "end if", only the entries from the original depth of 1.
+# CHECKERR: -> end if
+# CHECKERR: > end if
+
+# Test some invalid depth values. They should be ignored, and tracing should
+# proceed as if no depth was given, i.e. show everything.
+
+# Set depth to default first:
+set -e fish_trace_depth
```
Comment (Nahor) [inline tests/checks/trace.fish:210]: You should test that deleting actually works
```fish
set fish_trace_depth 1
set -e fish_trace_depth
{ { true } }
# CHECKERR: > set fish_trace_depth 1
# CHECKERR: > set -e fish_trace_depth
# CHECKERR: > begin
# CHECKERR: -> begin
# CHECKERR: --> true
# CHECKERR: -> end begin
# CHECKERR: > end begin
```

File: tests/checks/trace.fish [hunk 10]
```diff
@@ -65,6 +65,239 @@ end
...
+# CHECKERR: > set -e fish_trace_depth
+
+set fish_trace_depth a
```
Comment (Nahor) [inline tests/checks/trace.fish]: You need to fish_trace_depth to some other small value. Otherwise you don't know if the value was actually reset or if it just kept the old value (essentially same code as the test for deletion above).

(same for the other invalid values, of course)

File: src/env_dispatch.rs [hunk 11]
```diff
@@ -347,6 +348,25 @@ fn handle_fish_trace(vars: &EnvStack) {
...
+        crate::trace::trace_set_depth_default();
+        return;
+    };
+    let Ok(new_depth) = fish_wcstoi(&depth_str.as_string()) else {
+        flog!(env_dispatch, "Trace depth needs to be an integer");
+        crate::trace::trace_set_depth_default();
+        return;
+    };
+    let Ok(new_depth) = usize::try_from(new_depth) else {
+        flog!(env_dispatch, "Trace depth needs to be a positive integer");
```
Comment (Nahor) [inline src/env_dispatch.rs]: To answer your reply in the previous version, it could just say `invalid trace depth value` or something generic like that.
... which mean this code could be even further simplified:
```rust
fn handle_fish_trace_depth(vars: &EnvStack) {
    let Some(var) = vars.get(L!("fish_trace_depth")) else {
        crate::trace::trace_set_depth_default();
        return;
    };
    let Some(new_depth) = fish_wcstoi(&var.as_string())
        .ok()
        .and_then(|v| usize::try_from(v).ok())
    else {
        flog!(env_dispatch, "Invalid trace depth, using default");
        crate::trace::trace_set_depth_default();
        return;
    };
    crate::trace::trace_set_depth(new_depth);
}
```

A couple of other things I saw when trying it:
- `flog!` adds a space between its arguments, so there are 2 space in the log `Setting trace depth␣␣123`. `flogf!` would be better but it does not support `usize`, but...
- I don't know if we should log anything anyway. The user just asked for the change, we shouldn't need to tell them (unless there is an error of course).
- and it would be better to use `u32` for the depth value, this makes the max value the same regardless on the platform integer size.

File: tests/checks/trace.fish (hunk 5 above)
Comment (Nahor) [reply tests/checks/trace.fish:281]: I guess one would use something like `set -l fish_trace_depth (test "$i" -eq 2 && echo 10 || echo 0)` for your old example

File: tests/checks/trace.fish (hunk 8 above)
Comment (tjkirch) [reply tests/checks/trace.fish]: Yeah, that is nicer.  I kept newlines on the test of local scope to make it clearer when things are happening, but it definitely helped with all the invalid value tests.

File: src/env_dispatch.rs (hunk 11 above)
Comment (tjkirch) [reply src/env_dispatch.rs]: Thanks.  I changed all of that, except for using u32.  My reasoning is that you'd have to convert it into a usize anyway, to compare against parser.blocks_size, which is a fallible conversion, so it mucks up that section.  I rationalized the rest as benefiting 64-bit users with >4,294,967,296 trace depth levels if desired.  (kidding)  If it's better to have the same max regardless of platform, I'll add the conversions.

Comment (Nahor) [reply src/env_dispatch.rs]: u32 can convert to usize without fail.

**Comment by Nahor** [comment]
Code looks good. I don't have a particular opinion on the worthiness of the feature though, so I'll refrain from approving.

File: tests/checks/trace.fish (hunk 1 above)
Comment (krobelus) [reply tests/checks/trace.fish:87]: > I'm writing a script.

I think one alternative is `fish_trace=1 fish -c 'echo 123' &| grep -E '^-{0,2}>'`

> I'm running a number of commands interactively that I'm trying to understand better, and I don't have to think about how or where to add grep on each one.

Hmm maybe interactive shell+terminal is not an ideal environment for this.
Wouldn't it be better to pipe the entire `fish_trace` into an editor window, and then tell the editor to collapse at a given level.

Not sure yet. Do you have a concrete example where you find this useful?
I've never really filtered the entire output like this but I think I sometimes filter jump between matches of `^-->` in the editor or pager.

Comment (tjkirch) [reply tests/checks/trace.fish:87]: > Do you have a concrete example where you find this useful?

Sure.  I write a lot of scripts for myself - to manage my system, call other tools, etc.  I'm switching to fish because it's a nicer language than bash and I believe in where it's headed.  Because the scripts are for myself, I don't need fancy output, but I do want to know what's happening at various points.  `fish_trace` is perfect for this, but it shows way more than is helpful for this case.  I don't need to know the complete depth of expansion for every function; for example, the built-in `cd` function expands to 27 lines of trace, mostly at depths 2 and 3.  Setting a trap prints 41 lines down to depth 7.  If I can set depth to 0 or 1, I can see the commands I actually wrote, without so many implementation details.

> I think one alternative is `fish_trace=1 fish -c 'echo 123' &| grep -E '^-{0,2}>'`

Sure, and I'm sure there are other clever ways to do it as well, but if I have to wrap scripts to get useful output, it doesn't seem very friendly, and I'm probably not going to do it.

Comment (krobelus) [reply tests/checks/trace.fish:87]: Interesting. For simple scripts, I still use POSIX sh (where `set -x; cd` won't have noisy output) and for larger ones maybe Python.
Your explanation makes sense.
But this approach seems not quite perfect -- it will not work well if you use lots of `if`, `while` and other block statements.
Maybe what you really want is a feature to trace all commands in the current file?
Or if you use separate files for functions, all user-defined functions in ~/.config/fish/functions/ ?
Maybe something like `fish_trace=user` would make sense.

Comment (tjkirch) [reply tests/checks/trace.fish:87]: I agree that `fish_trace=user` would be nice.  I mentioned a similar idea in the linked issue, `fish_trace=exec`, which would only show external commands that fish runs.  I'm sure other filter types could be useful in different cases.

I think that _what_ gets traced is orthogonal to _how far_ it gets traced, and both are useful.  I've done the first (admittedly easier) part in this PR, which would solve my current issue, and I'd love to see trace filters in the future as well.
